### ISSUE TITLE
Is it possible to have different bindings for mouse key up and mouse key down?

### ISSUE THREAD

**Issue by KillaPL** [ruby2d/ruby2d#27]
I seem to be only able to do

```
on(mouse: 'any') do
  ...
end
```

Which is good and nice, but I would also be able to make a good use of being able to differentiate between mouse up and mouse down events. I'm not sure if I understand the implementation properly, but it seems that only one event is called.

Is it possible to have a distinction between different mouse events?

EDIT

Also I noticed left and right keys are treated the same, would it be possible to have this recognised?

**Comment by blacktm** [comment]
Anything is possible! 🌈

This was actually on my shortlist. On the native side (MRI, MRuby), we can use [SDL_MouseButtonEvent](https://wiki.libsdl.org/SDL_MouseButtonEvent), and on the web, looks like [MouseEvent.button](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/button) is the ticket. These get implemented in [Simple 2D](https://github.com/simple2d/simple2d) and [Simple2D.js](https://github.com/simple2d/simple2d.js), respectively. I can start a PR for those. Then on the Ruby 2D side, these dependencies get updated, along with the native and web extensions. And finally, we would update `window.rb`, test, and done.

**Comment by KillaPL** [comment]
Implementing those methods in c is probably above my skill, but if those features are available in c land, I can write ruby code to expose them : )

**Comment by blacktm** [comment]
PR for adding mouse events to Simple 2D: https://github.com/simple2d/simple2d/pull/50
Next up: implement the same in Simple2D.js, implement in Ruby 2D native and web extensions, write tests.

**Comment by blacktm** [comment]
PR for Simple2D.js: https://github.com/simple2d/simple2d.js/pull/3
This is on par in terms of functionality with the Simple 2D PR above ☝️

**Comment by KillaPL** [comment]
I was thinking about just a natural extension of what was earlier. First of all, `on mouse: "any"` should probably be deprecated, or become an alias to `on mouse: 'left'`. With mouse move and mouse scroll events, `any` is way to vague. I think this would be good:

```
on mouse: which, type: type do |x, y, type|
end
```
There `which` is one of `:left`, `:right`, `:move`, `:wheel` and whatever keys there are.
Passing additional type will only call the block if event that happend is of that type.

Types per `which` would probably match something like this:

:left, :right, :wheeldown and any button that can be pressed => `:down`, `:hold`, `:up`
:wheel probably needs `:up` and `:down`
:move probably doesn't need additional types.

I am thinking about having a code something like this:

```
class Button
  def initialize(opts = {})
    @on_click   = opts[:on_click]
    @on_hover   = opts[:on_hover]

    setup_events!
  end

  def contains?(x, y)
    # some logic to determine if given x and y coordinate are above the button
  end

  private

  def setup_events!
    if @on_click
      on mouse: :left, type: :down  do |x, y, type|
        if contains?(x, y)
          @on_click.call
        end
      end
    end

    if @on_hover
      on mouse: :move do |x, y|
        if contains?(x, y)
          @on_hover.call
        end
      end
    end
  end
end

Button.new({
  on_click: -> { },
  on_hover: -> { }
})
```

Namely, I want to set up menu logic and forget about it, and not have to check if mouse is over any of those buttons in update loop, like I do right now.
This poses a couple of interesting problems.

Currently, there can be only one `on mouse: "any"` event, so only the last button defined would have his action in place. For this idea to work, there would have to be an array of mouse callbacks for each type of events, and when their occur, this array would have to be traversed and each callback would have to be called.

If this is where it progressess, right away the `on` method must return something, that I would be able to call
`remove_callback(something_returned_by_on)` to remove the callback I set. For example when I hide the menu.

This would be significantly more complex, but significantly more powerful. Also, this would make it more similiar to how javascript window works. This is probably a matter of taste, but I personally think js works fine in that matter.

The question here is if you want to progress in this way. Certainly I, writing a clone of dwarf fortress, will probably have a rich menu and organising it like this seems reasonable for my needs.

As a bonus point, all of what I explained is mostly backwards compatible : )

I think I would try implementing them like this.

**Comment by KillaPL** [comment]
@blacktm Considering our discussion about `delta_x` and `delta_y`, there seems like there would be a lot of ideas going into the event handler. One idea would be to introduce `MouseEvent` class, and yield an instance of it the event handler, like so:

```
  on mouse: :left do |mouse_event|
  end
```

and it would respond to x, y, delta_x, delta_y, type and which, or whichever field names seem appriopriate.

**Comment by blacktm** [comment]
I've been digesting your comments — thanks for these. I definitely agree with you on simplifying the event param, and actually was thinking we could go further and just have an event name symbol. So, for example:

`on :mouse do |event|` — All mouse events
`on :mouse_down do |event|` — When mouse button pressed
`on :mouse_up do |event|` — When mouse button released
`on :mouse_move do |event|` — When mouse motion / movement
`on :mouse_scroll do |event|` — When mouse scrolling, wheel or trackpad

`on :key do |event|` — All keyboard events
`on :key_down do |event|` — When key is pressed, fired once
`on :key_held do |event|` — When key is being held down, fired every frame
`on :key_up do |event|` — When key released, fired once

`on :controller do |event|` — All controller events
`on :controller_axis do |event|` — When controller axis motion, like  analog sticks
`on :controller_button do |event|` — When controller button is pressed 

So, in your `setup_events!` example above, it would instead be:

```ruby
# on mouse down...
on :mouse do |event|
  if event.button == :left && event.type == :down
    @on_click.call if contains? event.x, event.y
  end
end

# on mouse move...
on :mouse_move do |e|
  @on_hover.call if contains? e.x, e.y
end
```

We _could_ even be more specific with the event symbol, although not sure how I feel about it:

```ruby
on :mouse_down_left do |event|
  @on_click.call if contains? event.x, event.y
end
```

[`Struct`](https://ruby-doc.org/core-2.2.0/Struct.html) is probably perfect for the `event`:

```ruby
MouseEvent = Struct.new(:type, :btn, :direction, :x, :y, :delta_x, :delta_y)
```

We'll have to allow many events to be registered (not overwritten like currently implemented). Not sure how adding/removing a callback would work, maybe something like JavaScript’s [`addEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) like you alluded to. Events could even be given a name. Might have to play around with some ideas.

**Comment by KillaPL** [comment]
I think list of APIs you suggested would work perfectly. As for the `on :mouse_down_left`, I'm not really sure if it's an overkill, or if this would be useful. I would have to get the API and play with it a little, see which evens seem naturally to me.

As for the events removing, I indeed thought of something like addEventListener, except doing things like

```
event_listener = on :mouse_down do 
end

remove_event(event_listener)
``` 

If there is a possibility to set separate callback for each event type like you mentioned, I could give a try on implementing having a list of them, and removing callbacks, since that would be ruby. I think I have an implementation in mind. Or I could provide a PR with this working on current `on mouse: "any"`, if you'd like to.

**Comment by blacktm** [comment]
Alrighty, just pushed a WIP PR, see #43. This adds all the mouse events we've talked about, including a refactor to `key` and `controller` as described above. Only difference is `:controller_button` was split into two events, `:controller_button_down` and  `:controller_button_up`. I also deleted `test/input.rb`, since it was useless, and added `test/mouse.rb` as a focused, interactive diagnostic test (it's pretty fun to play with).

This currently works in MRI and native — have to complete the web extension next. If you want to give this a try locally, you'll have to get the latest Simple 2D changes on `master`, either cloning and running `make rebuild` or using `simple2d install --HEAD` on linux or `brew install --HEAD simple2d` on macOS (can't remember which OS you're on).

I haven't started on the "events management" feature, but maybe we can get this merged first and add that as a second PR.
